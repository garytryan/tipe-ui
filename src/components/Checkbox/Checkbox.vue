<template>
  <div class="checkbox">
    <tipe-icon 
      :width="checkSize.width" 
      :height="checkSize.height" 
      class="check-icon" 
      icon="check" />
    <input 
      :class="classObject"
      :name="name"
      :data-tipe-ui="$options.name" 
      :disabled="isDisabled"
      :id="label"
      :value="value" 
      :tabindex="tabindex"
      :checked="checked"
      type="checkbox"
      @click="inputEmit('click')"
      @blur="inputEmit('blur')"
      @focus="inputEmit('focus')"
      @change="inputEmit('change')">
    <label 
      :for="label" 
      class="label" >{{ label }}</label>

  </div>
</template>

<script>
import vueTypes from 'vue-types'
import inputProps from '@/types/InputProps'
import TipeIcon from '../Icon'

export default {
  name: 'TipeCheckbox',
  components: { TipeIcon },
  props: {
    label: vueTypes.string,
    checked: vueTypes.bool.def(false),
    ...inputProps
  },
  computed: {
    classObject() {
      const { size, waiting, status } = this
      return {
        [size]: true,
        success: status === 'success',
        error: status === 'error',
        warning: status === 'warning',
        waiting
      }
    },
    isDisabled() {
      if (this.waiting || this.disabled) return true
      return false
    },
    checkSize() {
      switch (this.size) {
        case 'small':
          return {
            height: '6px',
            width: '8px'
          }
        case 'medium':
          return {
            height: '8px',
            width: '10px'
          }
        case 'large':
          return {
            height: '10px',
            width: '12px'
          }
        default:
          return {
            height: '8px',
            width: '10px'
          }
      }
    }
  },
  methods: {
    inputEmit(type) {
      this.$emit(type)
    }
  }
}
</script>

<style scoped lang="postcss">
.checkbox {
  display: grid;
  grid-template-rows: 1fr;
  grid-auto-columns: 1fr 1fr;
  grid-auto-flow: column;
  align-items: center;
  z-index: 3;
  justify-items: center;
}

.check-icon {
  grid-row: 1/1;
  grid-column: 1/1;
  z-index: 2;
}

input:disabled {
  cursor: not-allowed;
}

input {
  margin: 0;
  width: 0.9375rem;
  height: 0.9375rem;
  border-radius: 4px;
  border: solid 1px #b5baca;
  grid-row: 1/1;
  grid-column: 1/1;

  &.small {
    border-radius: 2px;
    width: 0.625rem;
    height: 0.625rem;
  }

  &.large {
    width: 1.125rem;
    height: 1.125rem;
  }
  &.success {
    border: solid 1px var(--success);
  }

  &.warning {
    border: solid 1px var(--warning);
  }

  &.error {
    border: solid 1px var(--error);
  }

  &.waiting {
    cursor: wait;
  }
}

input:checked {
  background-color: #6849fc;
  border: none;
  &.success {
    border: solid 1px var(--success);
  }

  &.warning {
    border: solid 1px var(--warning);
  }

  &.error {
    border: solid 1px var(--error);
  }
}

.label {
  letter-spacing: 0.3px;
  color: #1f346c;
}
</style>
