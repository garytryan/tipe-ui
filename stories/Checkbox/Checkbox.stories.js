import { storiesOf } from '@storybook/vue'

import TipeCheckbox from '../../src/components/Checkbox'

const style = () => ({
  width: '200px'
})

const props = () => ({
  label: 'Name',
  value: 'Olivia',
  status: 'success'
})
storiesOf('Checkbox', module).add('default checkbox', () => ({
  components: { TipeCheckbox },
  computed: { style, props },
  template: '<tipe-checkbox label="Name" />'
}))
