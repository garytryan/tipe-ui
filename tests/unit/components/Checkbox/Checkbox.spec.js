import { shallowMount } from '@vue/test-utils'
import { createRenderer } from 'vue-server-renderer'
import Checkbox from '@/components/Checkbox'
import tipeTestUtils from '~/tests/TipeTestUtils'

describe('Checkbox', () => {
  describe('<template>', () => {
    it('matches previous snapshot', () => {
      const renderer = createRenderer()
      const propsData = { label: 'Javascript' }
      const wrapper = shallowMount(Checkbox, { propsData })
      renderer.renderToString(wrapper.vm, (err, str) => {
        if (err) throw new Error(err)
        expect(str).toMatchSnapshot()
      })
    })

    it('has correct data-tipe-iu attibute', () => {
      const wrapper = shallowMount(Checkbox)

      expect(wrapper.attributes()['data-tipe-ui']).toBe('TipeCheckbox')
    })

    it('contains an input tag', () => {
      const wrapper = shallowMount(Checkbox)

      expect(wrapper.contains('input')).toEqual(true)
    })

    it('input type is checkbox', () => {
      const wrapper = shallowMount(Checkbox)

      expect(wrapper.find('input').attributes().type).toEqual('checkbox')
    })

    it('contains a label', () => {
      const wrapper = shallowMount(Checkbox)

      expect(wrapper.findAll('label')).toHaveLength(1)
    })
  })
  describe(':props', () => {
    it(':value - should pass value input element', () => {
      const propsData = { value: 'javascript' }
      const wrapper = shallowMount(Checkbox, { propsData })
      expect(wrapper.find('input').element.value).toEqual(`${propsData.value}`)
    })

    it(':name - should pass name input element', () => {
      const propsData = { name: 'Javascript' }
      const wrapper = shallowMount(Checkbox, { propsData })
      expect(wrapper.find('input').element.name).toEqual(`${propsData.name}`)
    })

    tipeTestUtils.test.input.size(Checkbox)

    tipeTestUtils.test.input.status(Checkbox)

    tipeTestUtils.test.input.disabled(Checkbox, wrapper =>
      wrapper.find('input')
    )

    it(':checked - should pass checked to input', () => {
      const propsData = { checked: true }
      const wrapper = shallowMount(Checkbox, { propsData })
      expect(wrapper.find('input').attributes().checked).toEqual(
        `${propsData.checked}`
      )
    })

    tipeTestUtils.test.input.waiting(Checkbox)

    it(':waiting(true) - has disabled attribute', () => {
      const propsData = { waiting: true }
      const wrapper = shallowMount(Checkbox, { propsData })
      expect(wrapper.find('input').attributes().disabled).toBeTruthy()
    })
  })

  describe('@events', () => {
    it('@focus - should emit', () => {
      const wrapper = shallowMount(Checkbox)
      wrapper.find('input').trigger('focus')
      expect(wrapper.emitted().focus).toBeTruthy()
    })

    it('@blur - should emit', () => {
      const wrapper = shallowMount(Checkbox)
      wrapper.find('input').trigger('blur')
      expect(wrapper.emitted().blur).toBeTruthy()
    })

    it('@click - should emit', () => {
      const wrapper = shallowMount(Checkbox)
      wrapper.find('input').trigger('click')
      expect(wrapper.emitted().click).toBeTruthy()
    })

    it('@change - should emit', () => {
      const wrapper = shallowMount(Checkbox)
      wrapper.find('input').trigger('change')
      expect(wrapper.emitted().change).toBeTruthy()
    })
  })
})
