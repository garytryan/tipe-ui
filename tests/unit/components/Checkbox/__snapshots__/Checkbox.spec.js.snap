// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox <template> matches previous snapshot 1`] = `
<div data-tipe-ui="TipeCheckbox" class="checkbox medium">
  <div class="check-flex">
    <div class="check-grid">
      <TipeIcon-stub class="check-icon"></TipeIcon-stub>
      <input name="" id="Javascript" tabindex="0" type="checkbox">
    </div>
  </div>
  <label for="Javascript" class="label">Javascript</label>
</div>
`;
